File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n
  | n <= 0          = Nothing
  | aliquotSum < n  = Just Deficient
  | aliquotSum == n = Just Perfect
  | aliquotSum > n  = Just Abundant
  where
    aliquotSum = sum factors
    factors = [d | d <- [1..n-1], mod n d == 0]

File: ruby/perfect-numbers/perfect_numbers.rb
module PerfectNumber
  CATEGORIES = %w{perfect abundant deficient}

  def self.classify(number)
    raise RuntimeError unless number.positive?
    CATEGORIES[aliquot_sum(number) <=> number]
  end

  class << self
  private

    def aliquot_sum(number)
      (1...number).select { |divisor| number.modulo(divisor).zero? }.sum
    end
  end
end
